### PR TITLE
Make stack clean be more thorough by default

### LINKED ISSUE #2278: `stack clean` should be documented better
This command is mentioned in the online docs only very briefly ("clean" appears two times on http://docs.haskellstack.org/en/stable/GUIDE/) but its semantics is not defined. 

I think "how to delete results from earlier builds/trigger a rebuild" is a basic question. I can understand that the one true answer is "you don't, because builds are deterministic", but sometimes they are not (see issue #2277).

The executable is more helpful

```
 stack clean --help
Usage: stack clean ([PACKAGE] | [--full]) [--help]
  Clean the local packages

Available options:
  PACKAGE                  If none specified, clean all local packages
  --full                   Delete all work directories (.stack-work by default)
```

but I still don't know what "clean a package" means.  

My suggestion is to (keep this output, but) add a paragraph to the (online) docs. In fact, I'd expect that I would find there: a list of all `stack` commands, and then for each of them, a description of what it does. Instead, the User Guide is a mixture of motivation and tutorial. That's useful, but it leaves some questions open. (Yes, I know it's hard to get people to write documentation ...)

### LINKED ISSUE #3863: stack clean should be more thorough by default
Several times I have been caught out by expecting `stack clean` to reset state and fix some transient failure, then finding out that only removing the .stack-work directories was sufficient. And I have only just now discovered `stack clean --full`. In the most recent example two of us spent several hours troubleshooting weird NFData version-skew-ish doctest failures on one machine, which `stack clean` did not fix. Somewhat related: #2278, #2936 and #3152.

We were using stack 1.6.3 (and recent HEAD in my case). I don't know how to reproduce the failure state. I expected `clean` to reset state and be pretty much equivalent to starting with  fresh clone of the project. Instead the doctest failures persisted until we tried a fresh clone (and then remembered to try manually removing .stack-work). clean's --full flag was not obvious.

### PR DESCRIPTION
**Pull request by vanceism7**
Fixes #3863 and potentially fixes #2278 

Please include the following checklist in your PR:

* [ ] Any changes that could be relevant to users have been recorded in the ChangeLog.md
* [x] The documentation has been updated, if necessary.

I tested out `stack clean` and `stack purge`. It seems like theyre doing what they should be. Also added a small spot in the guide.md explaining the commands a little more indepth.

### CODE REVIEW

File: doc/GUIDE.md [hunk 1]
```diff
@@ -280,6 +280,23 @@ As you can see from that path (and as emphasized earlier), the installation is
...
 even different GHC versions installed by stack.
 
+## Cleaning your project
+You can clean up build artifacts for your project using the `stack clean` and `stack purge`.
+
+### `stack clean`
+`stack clean` Deletes the local working directories containing compiler output. 
+(typically '.stack-work/dist' by default)
+
+Use `stack clean <specific-package>` to delete the output for a package specific-package only.
```
Comment (dbaynard) [inline doc/GUIDE.md]:
```suggestion
Use `stack clean <specific-package>` to delete the output for a package _specific-package_ only.
```

File: doc/GUIDE.md [hunk 2]
```diff
@@ -280,6 +280,23 @@ As you can see from that path (and as emphasized earlier), the installation is
...
+
+### `stack purge`
+`stack purge` Deletes the local stack working directory, including extra-deps, git dependencies and the compiler output. 
+It does not delete any snapshot packages, compilers or installed programs. -- This essentially
+reverts your project to a completely fresh state, as if it had never been built.
```
Comment (dbaynard) [inline doc/GUIDE.md]: This last bit could be more precise. It isn't obvious what a 'completely fresh state' is. I'd have to study that part of the code to offer something that could be written here instead.

I suspect the list I wrote in my comment on the issue may not be comprehensive, so I'll get somebody else to check.

File: src/Stack/Clean.hs [hunk 3]
```diff
@@ -54,12 +55,17 @@ dirsToDelete cleanOpts = do
 -- | Options for @stack clean@.
 data CleanOpts
     = CleanShallow [PackageName]
-    -- ^ Delete the "dist directories" as defined in 'Stack.Constants.distRelativeDir'
+    -- ^ Delete the "dist directories" as defined in 'Stack.Constants.Config.distRelativeDir'
```
Comment (dbaynard) [inline src/Stack/Clean.hs]: 😀

File: src/Stack/Options/CleanParser.hs [hunk 4]
```diff
@@ -19,4 +19,6 @@ cleanOptsParser = CleanShallow <$> packages <|> doFullClean
         flag'
             CleanFull
             (long "full" <>
-             help "Delete all work directories (.stack-work by default) in the project")
+             help "Delete the local stack working directory (.stack-work by default).")
```
Comment (dbaynard) [inline src/Stack/Options/CleanParser.hs]: Ah — this is a subtle change in meaning (probably my fault). It is possible for a project to have multiple .stack-work directories. The old description indicates all are deleted; this implies only one is deleted.

Comment (vanceism7) [reply src/Stack/Options/CleanParser.hs]: Oh my bad, I wasn't aware that you could have multiple .stack-work directories. Ill fix that

Comment (dbaynard) [reply src/Stack/Options/CleanParser.hs]: I'm not sure how stack actually behaves, mind.

File: doc/GUIDE.md [hunk 5]
```diff
@@ -280,6 +280,23 @@ As you can see from that path (and as emphasized earlier), the installation is
 placed to not interfere with any other GHC installation, whether system-wide or
 even different GHC versions installed by stack.
 
+## Cleaning your project
+You can clean up build artifacts for your project using the `stack clean` and `stack purge`.
```
Comment (mattaudesse) [inline doc/GUIDE.md]: Minor nitpick: I'd change this line to be either:
```suggestion
You can clean up build artifacts for your project using `stack clean` and `stack purge`.
```

or
```suggestion
You can clean up build artifacts for your project using the `stack clean` and `stack purge` commands.
```

Comment (mattaudesse) [reply doc/GUIDE.md]: Thanks for the PR @vanceism7 :+1:

**Comment by dbaynard** [comment]
It would be good to get an integration test, for this, too, but don't worry too much about it, @vanceism7, if you don't have a chance — it should be fairly straightforward.

**Comment by dbaynard** [comment]
Also I have a directory tree diff listing what has been deleted by each of these commands, which I still need to check. It's in progress! Thanks for bearing with me 👍

**Comment by vanceism7** [comment]
Oh no problem, thanks for taking the time to address this issue @dbaynard. Id be happy to put the integration test in, but I'll be a bit busy for the next week or so. If there's no rush, I'll try to get to that soon, otherwise, I'm fine if someone else finishes that up too.

Thanks!

**Comment by vanceism7** [comment]
Hey just posting an update, I'm back and finished with my prior obligations so I'll work on getting those integration tests in

**Comment by dbaynard** [comment]
We've had various issues with `stack clean` — I'll take a look at this shortly and see whether this can be merged. Thanks for your patience, @vanceism7!

**Comment by dbaynard** [comment]
I've rebased, and fixed a conflict (I thought there might be more). Given this doesn't actually affect the operation of clean (just the interface) if it passes CI & integration tests are ok then I'll merge this ASAP.

**Comment by dbaynard** [comment]
Note: the integration test fails. The behaviour of `stack clean` has changed on master, at some point — see #4480.

**Comment by vanceism7** [comment]
Oh ok np, I'll fix that integration test as soon as I can

**Comment by dbaynard** [comment]
@vanceism7 To be clear, it's not failing because of your changes! There's an issue with `clean` on `master`.

**Comment by snoyberg** [comment]
I've opened up #4607, which fixes the `clean` problems. I've pushed a new commit (https://github.com/commercialhaskell/stack/commit/1cabb987d4214a55fbd5a18a4788c4a4757bd562) to a separate branch, which merges this branch and the branch for #4607, together with a minor doc fix. Assuming #4607 is merged, I believe https://github.com/commercialhaskell/stack/commit/1cabb987d4214a55fbd5a18a4788c4a4757bd562 can be merged into this PR.

**Pull request merged**